base/iterators.jl

```diff
@@ -924,47 +926,97 @@ julia> collect(Iterators.partition([1,2,3,4,5], 2))
  [1, 2]
  [3, 4]
  [5]
+
+julia> collect(Iterators.partition([1,2,3,4,5], 2; flush=true))
+2-element Array{Array{Int64,1},1}:
+ [1, 2]
+ [3, 4]
+
+julia> collect(Iterators.partition([1,2,3,4,5], 2; step=1))
+2-element Array{Array{Int64,1},1}:
+ [1, 2]
+ [2, 3]
+ [3, 4]
+ [4, 5]
 ```
 """
-partition(c::T, n::Integer) where {T} = PartitionIterator{T}(c, Int(n))
-
+function partition(c::T, n::Integer;
+                   step::Integer = n, flush::Bool = true) where {T}
+    n < 1 && throw(ArgumentError("n (= $n) must be larger than 1"))
+    step < 1 && throw(ArgumentError("step (= $step) must be larger than 1"))
+    if T <: AbstractVector
+        return PartitionIterator{T,Nothing}(c, Int(n), Int(step), flush)
+    elseif step >= n
+        return PartitionIterator{T,:GE}(c, Int(n), Int(step), flush)
+    else
+        return PartitionIterator{T,:LESS}(c, Int(n), Int(step), flush)
+    end
+end
 
-struct PartitionIterator{T}
+struct PartitionIterator{T,STEP}
     c::T
     n::Int
+    step::Int
+    flush::Bool
 end
 
-eltype(::Type{PartitionIterator{T}}) where {T} = Vector{eltype(T)}
+eltype(::Type{<:PartitionIterator{T}}) where {T} = Vector{eltype(T)}
```

Comment:
It's unrelated to this PR, but I think the RHS `Vector{eltype(T)}` is problematic because we've been returning `view` from `partition` when iterating over a vector.  I suppose it has to reflect the actual type?